### ISSUE TITLE
TensorflowLite model can't run on Hexagon DSP device. Failed to apply delegate: Failed: Failed to execute graph.. ERROR: Node number 557 (TfLiteHexagonDelegate) failed to invoke.

### ISSUE THREAD

**Issue by sdkdzq1** [tensorflow/tensorflow#56747]
I convert the tfmodel to tflite, and when inference with python, it can run success. 
But in DSP, using the Hexagon C API (https://www.tensorflow.org/lite/performance/hexagon_delegate), in cpu run normal, but in DSP i get the below error. 
The hexagon_delegate so I used is https://storage.cloud.google.com/download.tensorflow.org/tflite/hexagon_nn_skel_v1.20.0.1.run
model only  include 2 conv and 1 conformer layers
Log
hexagon/ops/src/op_split.c:87:uneven split: 1 / 2
hexagon/src/execute.c:167:execute() failed on node id=181 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()

----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 139 (TfLiteHexagonDelegate) failed to invoke.

model can find here.
[encoder_6.zip](https://github.com/tensorflow/tensorflow/files/9098521/encoder_6.zip)

python code as this, and is running in tf2.4.0, the dsp's tf is 2.4.0 too.
import tensorflow as tf
import sys
import os
import numpy as np
tfmodel=sys.argv[1]#'temp.tflite'
wav_path='../../asr/BAC009S0764W0121.wav'
interpreter = tf.lite.Interpreter(model_path=tfmodel)
interpreter.allocate_tensors()
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
input_wav = np.random.rand(8000).astype(np.float32)
interpreter.set_tensor(input_details[0]['index'], input_wav)
interpreter.invoke()
output_data = interpreter.get_tensor(output_details[0]['index']).astype(np.float32)

**Comment by zhaoqing1992** [comment]
when i convert it to tflite, it has warnning: like this, WARNING:absl:Importing a function (__inference_inference_101322) with ops with custom gradients. Will likely fail if a gradient is requested.

convert code and saved model as this 
[Documents.zip](https://github.com/tensorflow/tensorflow/files/9098623/Documents.zip)

**Comment by zhaoqing1992** [comment]
I used the  conv + transfomer, to do the same work, it can run success.
so the error can find in some conv op
<img width="1163" alt="f6d50095-4e04-4e14-9932-1487eae1e1e9" src="https://user-images.githubusercontent.com/31097104/178682132-bad1323e-153e-4a1d-ba41-b743118d3b8a.png">

**Another thing**
the infer result is differ between CPU and DSP. as this
![image](https://user-images.githubusercontent.com/31097104/178682465-8464e20f-a2d3-4497-ae90-ea0ae0e003ce.png)

**Comment by mohantym** [comment]
Hi @zhaoqing1992 !
Could you confirm that your model follows  [8-bit symmetric quantization spec](https://www.tensorflow.org/lite/performance/quantization_spec). In case It finds any other unsupported op it will back to CPU. Can you give it a try in 2.8/2.9 version too?

 It would be handy to have device details too.

**Comment by zhaoqing1992** [comment]
1. The convert code uesd the Convert using float fallback quantization in https://www.tensorflow.org/lite/performance/post_training_integer_quant
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
convert.py is here https://github.com/tensorflow/tensorflow/files/9098623/Documents.zip

2. I have tried in 2.9, the error is diff
2.9 is 
`Log
hexagon/ops/src/op_split.c:87:uneven split: 1 / 2
hexagon/src/execute.c:167:execute() failed on node id=35 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()

----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 95 (TfLiteHexagonDelegate) failed to invoke.
`
2.4 is
`Log
hexagon/ops/src/op_split.c:87:uneven split: 1 / 2
hexagon/src/execute.c:167:execute() failed on node id=181 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()

----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 139 (TfLiteHexagonDelegate) failed to invoke.
`
3.  device details
device is the internal version
**/bigsur/bigsur:11/RQ3A.211001.001/1.130.72.32900-20220622:userdebug/test-keys

> Hi @zhaoqing1992 ! Could you confirm that your model follows [8-bit symmetric quantization spec](https://www.tensorflow.org/lite/performance/quantization_spec). In case It finds any other unsupported op it will back to CPU. Can you give it a try in 2.8/2.9 version too?
> 
> It would be handy to have device details too. Thank you!

**Comment by mohantym** [comment]
Ok @zhaoqing1992 ! Thanks for the update. 
@sachinprasadhs ! Could you look at this issue.
Thank you!

**Comment by sirakiin** [comment]
Hi @zhaoqing1992 

5dd4a61d6ed82f4f04f1d748d0c9e59a6e701b55 and 73cc8621b9c33424919d773f480d0b744c38c6c5 are in to fix the aforementioned issues in hexagon delegate.
You may pull the latest code and verify on your side.

Let me know if you have further questions.
Thanks,
Weiyi

**Comment by zhaoqing1992** [comment]
Thanks!

I have built the latest version, the tflite model can run on the device, but the CPU usage is very high, I confirmed it is running on the CPU. Which is not what I originally intended.

I expect the supported operators run on DSP and unsupported operators run on CPU, not all operators run on CPU.

> Hi @zhaoqing1992
> 
> [5dd4a61](https://github.com/tensorflow/tensorflow/commit/5dd4a61d6ed82f4f04f1d748d0c9e59a6e701b55) and [73cc862](https://github.com/tensorflow/tensorflow/commit/73cc8621b9c33424919d773f480d0b744c38c6c5) are in to fix the aforementioned issues in hexagon delegate. You may pull the latest code and verify on your side.
> 
> Let me know if you have further questions. Thanks, Weiyi

**Comment by sirakiin** [comment]
Hi @zhaoqing1992,

> I confirmed it is running on the CPU. Which is not what I originally intended.

Could you share the detailed logs, or the steps to reproduce it? Or you might use the [inference diff tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/evaluation/tasks/inference_diff) to reproduce.

For reference, I run the inference diff tool against your model (encoder_6.zip) with hexagon and shows most of the nodes are delegated to hexagon:

```
INFO: Initialized TensorFlow Lite runtime.
Func remote_handle_control not available on this device (NULL).
loaded libadsprpc.so
Hexagon delegate created.
INFO: TfLiteHexagonDelegate delegate: 136 nodes delegated out of 139 nodes with 2 partitions.

VERBOSE: Replacing 136 node(s) with delegate (TfLiteHexagonDelegate) node, yielding 5 partitions.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
I0000 00:00:1658183303.208241    8218 inference_profiler_stage.cc:78] Test interpreter has been initialized.
W0000 00:00:1658183303.209016    8218 tflite_inference_stage.cc:146]
VERBOSE: Replacing 12 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 7 partitions.
I0000 00:00:1658183303.209986    8218 inference_profiler_stage.cc:92] Reference interpreter (1 thread on CPU) has been initialized.
Num evaluation runs: 50
Reference run latency: avg=54294.8(us), std_dev=32215(us)
Test run latency: avg=11207.4(us), std_dev=2361(us)
OutputDiff[0]: avg_error=0.0214616, std_dev=0.00144275
```
The test interpreter is initialized with 136 nodes delegated out of 139 nodes by hexagon delegate.

Thanks!

**Comment by zhaoqing1992** [comment]
Hi, @sirakiin 
With the model(encoder_6.zip), the detailed logs as
`130|bigsur:/data/local/tmp # LD_LIBRARY_PATH=. ./test
WARNING: linker: Warning: "/data/local/tmp/libc++_shared.so" unused DT entry: unknown processor-specific (type 0x70000001 arg 0x0) (ignoring)
encoder.tflite 8000
INFO: Initialized TensorFlow Lite runtime.
INFO: Interpreter::UseNNAPI() is deprecated. Use tflite::NnApiDelegate() directly instead.
loaded libcdsprpc.so
INFO: TfLiteHexagonDelegate delegate: 137 nodes delegated out of 139 nodes with 1 partitions.

encoder.tflite
----------------
Timestamp: Mon Jan  1 17:57:01 2001
Log
hexagon/ops/src/op_split.c:87:uneven split: 1 / 2
hexagon/src/execute.c:167:execute() failed on node id=181 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()
----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 139 (TfLiteHexagonDelegate) failed to invoke.
`

The loaded so is libcdsprpc.so, which different with yours libadsprpc.so. And the device is from qualcomm.

My Hexagon libraries is generated from https://storage.cloud.google.com/download.tensorflow.org/tflite/hexagon_nn_skel_v1.20.0.1.run, is it same.

**Comment by zhaoqing1992** [comment]
I have used the [inference diff tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/evaluation/tasks/inference_diff) to reproduce. The result is same to your log.
Thanks!

But when the number of Conformer layers is four, a new error is happened. 
Model is here.
[tmp.aa.zip](https://github.com/tensorflow/tensorflow/files/9156461/tmp.aa.zip)
[tmp.ab.zip](https://github.com/tensorflow/tensorflow/files/9156464/tmp.ab.zip)


'''
./run_eval  --model_file=encoder.12.con.4l.tflite --delegate=hexagon                                                                                           
INFO: Initialized TensorFlow Lite runtime.
loaded libcdsprpc.so
Hexagon delegate created.
INFO: TfLiteHexagonDelegate delegate: 505 nodes delegated out of 532 nodes with 5 partitions.

VERBOSE: Replacing 505 node(s) with delegate (TfLiteHexagonDelegate) node, yielding 9 partitions.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
native : lite/tools/evaluation/stages/inference_profiler_stage.cc:78 Test interpreter has been initialized.
native : lite/tools/evaluation/stages/tflite_inference_stage.cc:146 
VERBOSE: Replacing 48 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 25 partitions.
native : lite/tools/evaluation/stages/inference_profiler_stage.cc:92 Reference interpreter (1 thread on CPU) has been initialized.
----------------
Timestamp: Mon Jan  1 10:12:03 2001


Log
hexagon/ops/src/op_softmax_d32.c:214:softmax_32: range = -203399.890625 ... 192531.187500, beta = 1.000000: beta too large
hexagon/src/execute.c:167:execute() failed on node id=f7 err=-1
hexagon/src/interface.c:1297:fail in execute_inner()

----------------
ERROR: Failed: Failed to execute graph..
ERROR: Node number 532 (TfLiteHexagonDelegate) failed to invoke.
native : lite/tools/evaluation/stages/tflite_inference_stage.cc:182 TFLite interpreter failed to invoke at run 0
Could not run the task evaluation!
'''

Could you give me a hand.
Thanks!

**Comment by sirakiin** [comment]
Hi @sdkdzq1 sorry for the delay.

For this model, the issue comes from the softmax node with a too large quantization range.

See hexagon [nnlib](https://source.codeaurora.org/quic/hexagon_nn/nnlib/tree/hexagon/ops/src/op_softmax_d32.c#n115) for the restrictions. Could you check your model quantization and see if it's expected?

**Comment by sdkdzq1** [comment]
Hi @sirakiin,
I have trained several different layers of the conformer model, and such errors will not occur in layers 1, 2 and 3.
Therefore, I use netron to check the model quantization of the fourth softmax. The message as blow,
![image](https://user-images.githubusercontent.com/31097104/181694374-223788ac-0c25-4f00-9bc7-7ab70ffcca36.png)

I guess the error may be caused by excessive cumulative error of TfLiteHexagonDelegate.

The convert code uesd the Convert using float fallback quantization in https://www.tensorflow.org/lite/performance/post_training_integer_quant,
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.representative_dataset is the feature extraction from training data.
